sdk/core/azure-core/azure/core/pipeline/transport/_requests_basic.py

```diff
@@ -130,17 +130,25 @@ def __next__(self):
                 self.response.internal_response.close()
                 raise StopIteration()
             except (requests.exceptions.ChunkedEncodingError,
-                    requests.exceptions.ConnectionError):
+                    requests.exceptions.ConnectionError) as ex:
                 retry_total -= 1
                 if retry_total <= 0:
                     retry_active = False
                 else:
                     time.sleep(retry_interval)
                     headers = {'range': 'bytes=' + str(self.downloaded) + '-'}
-                    resp = self.pipeline.run(self.request, stream=True, headers=headers)
-                    if resp.http_response.status_code == 416:
-                        raise
-                    chunk = next(self.iter_content_func)
+                    try:
+                        resp = self.pipeline.run(self.request, stream=True, headers=headers)
+                        if not resp.http_response:
+                            raise
+                        if resp.http_response.status_code == 416:
+                            raise
+                        self.iter_content_func = resp.http_response.iter_content(self.block_size)
```

Comment:
This last change of re-setting `self.iter_content_func` seems to make sense superficially, because now `next(self.iter_content_func)` but it really does not help for the reasons I explained already in some depth, so I'm just citing myself here:
```
And as already commented, the code makes assumptions it cannot safely make: StreamDownloader is also used for requests that already have a (non-trivial) range header set. So what happens if the original range header is, say, "bytes=1000-1999", chunking is 200 and there is a connection error after the first chunk?
This code will then go on a make a request for "range: bytes=200-", which does not overlap with the original request.
```

The only way out I can see here is to remove the retry entirely. This function just does not have enough context information to re-try safely. So the best it can do is to raise and let the upper layers handle it.

Reply:
Thank you.

Yes. I was about to answer this. :)

As you know, the purpose for azure-core is to support other SDK libraries. We have not heard any real scenarios from libraries that need to read bytes like "bytes=1000-1999". In storage, we always download entire file. This made us leaning towards having retry rather than failing directly. 

But yes, you are right. It is not correct in case "bytes=1000-1999". I will open an issue to track it.

In the meantime, do you have real scenarios that need this?

That can help us prioritize the issue.

Thanks.

Reply:
Surely you are aware this is used in the azure-storage-blob library, which *does* set range headers by default, as it will use multi-threading to download large blobs in chunks, in parallel?

Yes, we have real scenarios we use this. As you might be aware, we use azure storage blob for big data analysis, processing dozens of TB daily. We had several customer-facing incidents that are very likely due to this bug. We spent person-weeks tracking this down and working around this.